text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1045 @@
+
+Thus we have a marker (`const`) to add in front of functions that requires the function body to only contain things const eval can handle. This in turn allows a `const` annotated function to be called from const contexts, as you now have a guarantee it will stay callable.
+
+When calling a trait method, this simple scheme (that works great for free functions and inherent methods) does not work.
+
+Throughout this document, we'll be revisiting the example below. Method syntax and `dyn Trait` problems all also exist with static method calls, so we'll stick with the latter to have the simplest examples possible.
+
+```rust
+const fn default<T: Default>() -> T {
+    T::default()
+}
+
+// Could also be `const fn`, but that's an orthogonal change
+fn compile_time_default<T: Default>() -> T {
+    const { T::default() }
+}
+```
+
+Neither of the above should (or do) compile.
+The first, because you could pass any type T whose default impl could
+
+* mutate a global static,
+* read from a file, or
+* just allocate memory,
+
+which are all not possible right now in const code, and some can't be done in Rust in const code at all.
+
+It should be possible to write `default` in a way that allows it to be called in const contexts
+for types whose `Default` impl's `default` method satisfies all rules that `const fn` must satisfy
+(including some annotation that guarantees this won't break by accident).
+It must always be possible to call `default` outside of const contexts with no limitations on the generic parameters that may be passed.
+
+Similarly it should be possible to write `compile_time_default` in a way that also requires calls
+outside of const contexts to only pass generic parameters whose `Default::default` method satisifies
+the usual `const fn` rules. This is necessary in order to allow a const block
+(which can access generic parameters) in the function body to invoke methods on the generic parameter.
+
+So, we need some annotation that differentiates a `T: Default` bound from one that gives us the guarantees we're looking for.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Nomenclature and new syntax concepts
+
+### Const trait methods
+
+Traits can declare methods as `const`. Doing so is a breaking change, as all impls are now required to provide a `const` method,
+which existing impls can't.
+
+```rust
+trait Trait {
+    const fn method();
+}
+```
+
+These methods need to be implemented as `const`:
+
+```rust
+impl Trait for Type {
+    const fn method() {}
+}
+```
+
+### Const trait bounds
+
+Any item that can have trait bounds can also have `const Trait` bounds.
+
+Examples:
+
+* `T: const Trait`, requiring any type that `T` is instantiated with to have a trait impl with `const` methods for `Trait`.
+* `dyn const Trait`, requiring any type that is unsized to this dyn trait to have a trait impl with `const` methods for `Trait`.
```

Comment:
Just to be sure, this means that *all* methods (those declared `fn` and `const fn` in `Trait`) must be `const` in the `impl`?